File: frontend/js/pages/relatorios.js
import { api } from '../api.js';

function fmtDt(s) { return s ? new Date(s).toLocaleString('pt-BR') : '—'; }

function section(id, title, bodyHtml) {
  return `<div class="section-card" id="sec-${id}">
    <div class="section-header" data-sec="${id}">
      <h3>${title}</h3>
      <span class="section-toggle" id="toggle-${id}">▼</span>
    </div>
    <div class="section-body" id="body-${id}" style="display:none">
      ${bodyHtml}
    </div>
  </div>`;
}

function table(cols, rows) {
  if (!rows.length) return '<p class="empty-state">Nenhum registro encontrado.</p>';
  return `<div class="table-wrap"><table>
    <thead><tr>${cols.map(c => `<th>${c}</th>`).join('')}</tr></thead>
    <tbody>${rows}</tbody>
  </table></div>`;
}

export async function mount(container) {
  let eventos = [], habilidades = [];

  async function loadAux() {
    [eventos, habilidades] = await Promise.all([api.eventos.list(), api.habilidades.list()]);
  }

  function eventoLabel(e) { return `#${e.id_evento} — ${e.tipo_evento?.descricao || '?'}`; }

  function eventoSel(id, selId = '') {
    return `<select id="${id}" style="padding:7px;background:var(--surface-2);border:1px solid var(--border);border-radius:var(--radius);color:var(--text);min-width:200px">
      <option value="">Selecione...</option>
      ${eventos.map(e => `<option value="${e.id_evento}" ${e.id_evento == selId ? 'selected' : ''}>${eventoLabel(e)}</option>`).join('')}
    </select>`;
  }

  function habSel(id) {
    return `<select id="${id}" style="padding:7px;background:var(--surface-2);border:1px solid var(--border);border-radius:var(--radius);color:var(--text);min-width:160px">
      <option value="">Selecione...</option>
      ${habilidades.map(h => `<option value="${h.id_habilidade}">${h.descricao}</option>`).join('')}
    </select>`;
  }

  // ── Render helpers ────────────────────────────────────────────────────────

  function renderEscala(data) {
    const rows = data.map(r => `<tr><td>${r.nome_pessoa}</td><td>${r.cpf_pessoa}</td><td>${r.habilidade}</td><td>${r.ministerio}</td></tr>`).join('');
    document.getElementById('result-escala').innerHTML = table(['Pessoa', 'CPF', 'Habilidade', 'Ministério'], data.length ? [rows] : []);
  }

  function renderAgenda(data) {
    const rows = data.map(r => `<tr>
      <td>${r.id_evento}</td><td>${r.tipo_evento}</td><td>${r.local}</td>
      <td>${fmtDt(r.dt_hr_prog_inicio)}</td><td>${fmtDt(r.dt_hr_prog_fim)}</td>
      <td>${r.habilidade}</td><td>${r.ministerio}</td>
    </tr>`).join('');
    document.getElementById('result-agenda').innerHTML = table(['ID', 'Tipo', 'Local', 'Início', 'Fim', 'Habilidade', 'Ministério'], data.length ? [rows] : []);
  }

  function renderParticipacao(data) {
    const rows = data.map(r => `<tr><td>${r.cpf}</td><td>${r.nome}</td><td>${r.total_alocacoes}</td></tr>`).join('');
    document.getElementById('result-participacao').innerHTML = table(['CPF', 'Nome', 'Total Alocações'], data.length ? [rows] : []);
  }

  function renderQuorum(data) {
    const rows = data.map(r => `<tr>
      <td>${r.habilidade}</td>
      <td>${r.qtd_necessaria}</td>
      <td>${r.qtd_alocada}</td>
      <td><span class="badge ${r.coberto ? 'badge-success' : 'badge-danger'}">${r.coberto ? 'Coberto' : 'Descoberto'}</span></td>
    </tr>`).join('');
    document.getElementById('result-quorum').innerHTML = table(['Habilidade', 'Necessário', 'Alocado', 'Status'], data.length ? [rows] : []);
  }

  function renderDistribuicao(data) {
    if (!data.length) { document.getElementById('result-dist').innerHTML = '<p class="empty-state">Nenhum registro.</p>'; return; }
    const max = Math.max(...data.map(d => d.total_participacoes), 1);
    const bars = data.map(d => `<div class="bar-row">
      <div class="bar-label" title="${d.ministerio}">${d.ministerio}</div>
      <div class="bar-track"><div class="bar-fill" style="width:${Math.round(d.total_participacoes/max*100)}%"></div></div>
      <div class="bar-value">${d.total_participacoes}</div>
    </div>`).join('');
    document.getElementById('result-dist').innerHTML = `<div class="bar-list">${bars}</div>`;
  }

  function renderPastores(data) {
    const rows = data.map(r => `<tr>
      <td>${r.cpf_visitante}</td><td>${r.nome_visitante}</td>
      <td>${r.cpf_convidador || '—'}</td><td>${r.nome_convidador || '—'}</td>
    </tr>`).join('');
    document.getElementById('result-pastores').innerHTML = table(['CPF Visitante', 'Visitante', 'CPF Convidador', 'Convidado Por'], data.length ? [rows] : []);
  }

  function renderOcupacao(data) {
    const rows = data.map(r => `<tr><td>${r.id_local}</td><td>${r.nome_local}</td><td>${r.total_eventos}</td></tr>`).join('');
    document.getElementById('result-ocupacao').innerHTML = table(['ID', 'Local', 'Total Eventos'], data.length ? [rows] : []);
  }

  function renderAptos(data) {
    const rows = data.map(r => `<tr><td>${r.cpf}</td><td>${r.nome}</td></tr>`).join('');
    document.getElementById('result-aptos').innerHTML = table(['CPF', 'Nome'], data.length ? [rows] : []);
  }

  // ── Build page ────────────────────────────────────────────────────────────

  container.innerHTML = `<div class="page-header"><h1>📈 Relatórios</h1></div>` + [

    section('escala', '1. Escala por Evento', `
      <div class="section-controls">
        <label>Evento<div id="escala-sel-wrap"></div></label>
        <button class="btn btn-primary btn-sm" id="btn-escala">Gerar</button>
      </div>
      <div id="result-escala"></div>
    `),

    section('agenda', '2. Agenda de Voluntário', `
      <div class="section-controls">
        <label>CPF da Pessoa<input type="text" id="inp-cpf-agenda" placeholder="11 dígitos" maxlength="11" style="min-width:140px"></label>
        <button class="btn btn-primary btn-sm" id="btn-agenda">Buscar</button>
      </div>
      <div id="result-agenda"></div>
    `),

    section('participacao', '3. Participação por Período', `
      <div class="section-controls">
        <label>De<input type="date" id="inp-ini-part"></label>
        <label>Até<input type="date" id="inp-fim-part"></label>
        <button class="btn btn-primary btn-sm" id="btn-participacao">Gerar</button>
      </div>
      <div id="result-participacao"></div>
    `),

    section('quorum', '4. Cobertura de Quórum', `
      <div class="section-controls">
        <label>Evento<div id="quorum-sel-wrap"></div></label>
        <button class="btn btn-primary btn-sm" id="btn-quorum">Verificar</button>
      </div>
      <div id="result-quorum"></div>
    `),

    section('dist', '5. Distribuição por Ministério', `
      <div class="section-controls">
        <button class="btn btn-primary btn-sm" id="btn-dist">Carregar</button>
      </div>
      <div id="result-dist"></div>
    `),

    section('pastores', '6. Visitantes e Convidadores', `
      <div class="section-controls">
        <button class="btn btn-primary btn-sm" id="btn-pastores">Carregar</button>
      </div>
      <div id="result-pastores"></div>
    `),

    section('ocupacao', '7. Ocupação de Locais', `
      <div class="section-controls">
        <button class="btn btn-primary btn-sm" id="btn-ocupacao">Carregar</button>
      </div>
      <div id="result-ocupacao"></div>
    `),

    section('aptos', '8. Voluntários Aptos Não Escalados', `
      <div class="section-controls">
        <label>Evento<div id="aptos-ev-wrap"></div></label>
        <label>Habilidade<div id="aptos-hab-wrap"></div></label>
        <button class="btn btn-primary btn-sm" id="btn-aptos">Buscar</button>
      </div>
      <div id="result-aptos"></div>
    `),

  ].join('');

  await loadAux();

  // Inject selects
  document.getElementById('escala-sel-wrap').innerHTML = eventoSel('sel-escala-ev');
  document.getElementById('quorum-sel-wrap').innerHTML = eventoSel('sel-quorum-ev');
  document.getElementById('aptos-ev-wrap').innerHTML = eventoSel('sel-aptos-ev');
  document.getElementById('aptos-hab-wrap').innerHTML = habSel('sel-aptos-hab');

  // Toggle collapse
  document.querySelectorAll('.section-header').forEach(header => {
    header.addEventListener('click', () => {
      const sec = header.dataset.sec;
      const body = document.getElementById(`body-${sec}`);
      const toggle = document.getElementById(`toggle-${sec}`);
      const open = body.style.display !== 'none';
      body.style.display = open ? 'none' : 'block';
      toggle.classList.toggle('open', !open);
    });
  });

  // Button handlers
  document.getElementById('btn-escala').onclick = async () => {
    const id = document.getElementById('sel-escala-ev').value;
    if (!id) { window.toast('Selecione um evento', 'error'); return; }
    try { renderEscala(await api.relatorios.escalaEvento(id)); }
    catch (err) { window.toast(err.message, 'error'); }
  };

  document.getElementById('btn-agenda').onclick = async () => {
    const cpf = document.getElementById('inp-cpf-agenda').value.trim();
    if (!cpf) { window.toast('Informe o CPF', 'error'); return; }
    try { renderAgenda(await api.relatorios.agendaPessoa(cpf)); }
    catch (err) { window.toast(err.message, 'error'); }
  };

  document.getElementById('btn-participacao').onclick = async () => {
    const ini = document.getElementById('inp-ini-part').value;
    const fim = document.getElementById('inp-fim-part').value;
    if (!ini || !fim) { window.toast('Informe o período', 'error'); return; }
    try { renderParticipacao(await api.relatorios.participacao(ini, fim)); }
    catch (err) { window.toast(err.message, 'error'); }
  };

  document.getElementById('btn-quorum').onclick = async () => {
    const id = document.getElementById('sel-quorum-ev').value;
    if (!id) { window.toast('Selecione um evento', 'error'); return; }
    try { renderQuorum(await api.relatorios.coberturaQuorum(id)); }
    catch (err) { window.toast(err.message, 'error'); }
  };

  document.getElementById('btn-dist').onclick = async () => {
    try { renderDistribuicao(await api.relatorios.distribuicaoMinisterio()); }
    catch (err) { window.toast(err.message, 'error'); }
  };

  document.getElementById('btn-pastores').onclick = async () => {
    try { renderPastores(await api.relatorios.pastoresVisitantes()); }
    catch (err) { window.toast(err.message, 'error'); }
  };

  document.getElementById('btn-ocupacao').onclick = async () => {
    try { renderOcupacao(await api.relatorios.ocupacaoLocais()); }
    catch (err) { window.toast(err.message, 'error'); }
  };

  document.getElementById('btn-aptos').onclick = async () => {
    const ev = document.getElementById('sel-aptos-ev').value;
    const hab = document.getElementById('sel-aptos-hab').value;
    if (!ev || !hab) { window.toast('Selecione evento e habilidade', 'error'); return; }
    try { renderAptos(await api.relatorios.voluntariosAptos(ev, hab)); }
    catch (err) { window.toast(err.message, 'error'); }
  };
}

File: backend/app/routers/relatorios.py
import datetime
from typing import List

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import text
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas import (
    AgendaItem,
    AptidaoHabilidadeItem,
    AtrasoEventoItem,
    CoberturaQuorumItem,
    EscalaEventoItem,
    OcupacaoLocalItem,
    ParticipacaoItem,
    PastorVisitanteItem,
    VoluntarioAptoItem,
)
from app.security import get_current_pessoa

router = APIRouter()


@router.get("/escala-evento/{id_evento}", response_model=List[EscalaEventoItem])
def escala_evento(
    id_evento: int,
    db: Session = Depends(get_db),
    _auth=Depends(get_current_pessoa),
):
    sql = text("""
        SELECT
            p.id_pessoa,
            p.nome AS nome_pessoa,
            h.descricao AS habilidade,
            CASE WHEN m.id_membro IS NOT NULL THEN 'membro' ELSE 'visitante' END AS tipo_pessoa
        FROM alocacao a
        JOIN pessoa p ON p.id_pessoa = a.id_pessoa
        JOIN habilidade h ON h.id_habilidade = a.id_habilidade
        LEFT JOIN membro m ON m.id_pessoa = p.id_pessoa
        WHERE a.id_evento = :id_evento
        ORDER BY h.descricao, p.nome
    """)
    rows = db.execute(sql, {"id_evento": id_evento}).mappings().all()
    return [EscalaEventoItem(**r) for r in rows]


@router.get("/agenda-pessoa/{id_pessoa}", response_model=List[AgendaItem])
def agenda_pessoa(
    id_pessoa: int,
    db: Session = Depends(get_db),
    _auth=Depends(get_current_pessoa),
):
    sql = text("""
        SELECT
            e.id_evento,
            te.descricao AS tipo_evento,
            l.nome AS local,
            e.dt_hr_prog_inicio,
            e.dt_hr_prog_fim,
            h.descricao AS habilidade
        FROM alocacao a
        JOIN evento e ON e.id_evento = a.id_evento
        JOIN tipo_evento te ON te.id_tipo_evento = e.id_tipo_evento
        JOIN local l ON l.id_local = e.id_local
        JOIN habilidade h ON h.id_habilidade = a.id_habilidade
        WHERE a.id_pessoa = :id_pessoa
          AND e.dt_hr_prog_inicio >= NOW()
        ORDER BY e.dt_hr_prog_inicio
    """)
    rows = db.execute(sql, {"id_pessoa": id_pessoa}).mappings().all()
    return [AgendaItem(**r) for r in rows]


@router.get("/participacao", response_model=List[ParticipacaoItem])
def participacao_periodo(
    ini: datetime.datetime = Query(..., description="Data/hora início do período"),
    fim: datetime.datetime = Query(..., description="Data/hora fim do período"),
    db: Session = Depends(get_db),
    _auth=Depends(get_current_pessoa),
):
    sql = text("""
        SELECT p.id_pessoa, p.nome, COUNT(*) AS total_alocacoes
        FROM pessoa p
        JOIN alocacao a ON a.id_pessoa = p.id_pessoa
        JOIN evento e ON e.id_evento = a.id_evento
        WHERE e.dt_hr_prog_inicio BETWEEN :ini AND :fim
        GROUP BY p.id_pessoa, p.nome
        ORDER BY total_alocacoes DESC
    """)
    rows = db.execute(sql, {"ini": ini, "fim": fim}).mappings().all()
    return [ParticipacaoItem(**r) for r in rows]


@router.get("/cobertura-quorum/{id_evento}", response_model=List[CoberturaQuorumItem])
def cobertura_quorum(
    id_evento: int,
    db: Session = Depends(get_db),
    _auth=Depends(get_current_pessoa),
):
    sql = text("""
        SELECT
            h.descricao AS habilidade,
            n.qtd AS qtd_necessaria,
            COUNT(a.id_pessoa) AS qtd_alocada,
            (COUNT(a.id_pessoa) >= n.qtd) AS coberto
        FROM necessita n
        JOIN habilidade h ON h.id_habilidade = n.id_habilidade
        JOIN evento e ON e.id_tipo_evento = n.id_tipo_evento
        LEFT JOIN alocacao a ON a.id_evento = e.id_evento AND a.id_habilidade = n.id_habilidade
        WHERE e.id_evento = :id_evento
        GROUP BY h.descricao, n.qtd
        ORDER BY coberto, h.descricao
    """)
    rows = db.execute(sql, {"id_evento": id_evento}).mappings().all()
    return [CoberturaQuorumItem(**r) for r in rows]


@router.get("/aptidao-habilidade", response_model=List[AptidaoHabilidadeItem])
def aptidao_habilidade(
    db: Session = Depends(get_db),
    _auth=Depends(get_current_pessoa),
):
    sql = text("""
        SELECT
            h.descricao AS habilidade,
            COUNT(DISTINCT po.id_pessoa) AS total_aptos,
            COUNT(a.id_pessoa) AS total_acionamentos
        FROM habilidade h
        LEFT JOIN possui po ON po.id_habilidade = h.id_habilidade
        LEFT JOIN alocacao a ON a.id_habilidade = h.id_habilidade
        GROUP BY h.descricao
        ORDER BY total_acionamentos DESC
    """)
    rows = db.execute(sql).mappings().all()
    return [AptidaoHabilidadeItem(**r) for r in rows]


@router.get("/pastores-visitantes", response_model=List[PastorVisitanteItem])
def pastores_visitantes(
    db: Session = Depends(get_db),
    _auth=Depends(get_current_pessoa),
):
    sql = text("""
        SELECT
            v.id_visitante,
            p.nome AS nome_visitante,
            v.e_pastor,
            v.convidado_por AS id_convidador,
            p2.nome AS nome_convidador
        FROM visitante v
        JOIN pessoa p ON p.id_pessoa = v.id_pessoa
        LEFT JOIN pessoa p2 ON p2.id_pessoa = v.convidado_por
        ORDER BY p.nome
    """)
    rows = db.execute(sql).mappings().all()
    return [PastorVisitanteItem(**r) for r in rows]


@router.get("/ocupacao-locais", response_model=List[OcupacaoLocalItem])
def ocupacao_locais(
    db: Session = Depends(get_db),
    _auth=Depends(get_current_pessoa),
):
    sql = text("""
        SELECT l.id_local, l.nome AS nome_local, COUNT(e.id_evento) AS total_eventos
        FROM local l
        LEFT JOIN evento e ON e.id_local = l.id_local
        GROUP BY l.id_local, l.nome
        ORDER BY total_eventos DESC
    """)
    rows = db.execute(sql).mappings().all()
    return [OcupacaoLocalItem(**r) for r in rows]


@router.get("/voluntarios-aptos", response_model=List[VoluntarioAptoItem])
def voluntarios_aptos(
    id_evento: int = Query(..., description="ID do evento"),
    id_habilidade: int = Query(..., description="ID da habilidade"),
    db: Session = Depends(get_db),
    _auth=Depends(get_current_pessoa),
):
    sql = text("""
        SELECT p.id_pessoa, p.nome
        FROM pessoa p
        JOIN possui po ON po.id_pessoa = p.id_pessoa
        WHERE po.id_habilidade = :id_habilidade
          AND p.id_pessoa NOT IN (
              SELECT a.id_pessoa FROM alocacao a WHERE a.id_evento = :id_evento
          )
        ORDER BY p.nome
    """)
    rows = db.execute(sql, {"id_habilidade": id_habilidade, "id_evento": id_evento}).mappings().all()
    return [VoluntarioAptoItem(**r) for r in rows]


@router.get("/controle-atrasos", response_model=List[AtrasoEventoItem])
def controle_atrasos(
    db: Session = Depends(get_db),
    _auth=Depends(get_current_pessoa),
):
    sql = text("""
        SELECT
            e.id_evento,
            te.descricao AS tipo_evento,
            e.dt_hr_prog_inicio,
            e.dt_hr_prog_fim,
            e.dt_hr_efet_inicio,
            e.dt_hr_efet_fim,
            CASE
                WHEN e.dt_hr_efet_inicio IS NOT NULL
                THEN EXTRACT(EPOCH FROM (e.dt_hr_efet_inicio - e.dt_hr_prog_inicio)) / 60
                ELSE NULL
            END AS atraso_inicio_min,
            EXTRACT(EPOCH FROM (e.dt_hr_prog_fim - e.dt_hr_prog_inicio)) / 60 AS duracao_planejada_min,
            CASE
                WHEN e.dt_hr_efet_inicio IS NOT NULL AND e.dt_hr_efet_fim IS NOT NULL
                THEN EXTRACT(EPOCH FROM (e.dt_hr_efet_fim - e.dt_hr_efet_inicio)) / 60
                ELSE NULL
            END AS duracao_efetiva_min
        FROM evento e
        JOIN tipo_evento te ON te.id_tipo_evento = e.id_tipo_evento
        ORDER BY e.dt_hr_prog_inicio DESC
    """)
    rows = db.execute(sql).mappings().all()
    result = []
    for r in rows:
        result.append(AtrasoEventoItem(
            id_evento=r["id_evento"],
            tipo_evento=r["tipo_evento"],
            dt_hr_prog_inicio=r["dt_hr_prog_inicio"],
            dt_hr_prog_fim=r["dt_hr_prog_fim"],
            dt_hr_efet_inicio=r["dt_hr_efet_inicio"],
            dt_hr_efet_fim=r["dt_hr_efet_fim"],
            atraso_inicio_min=int(r["atraso_inicio_min"]) if r["atraso_inicio_min"] is not None else None,
            duracao_planejada_min=int(r["duracao_planejada_min"]),
            duracao_efetiva_min=int(r["duracao_efetiva_min"]) if r["duracao_efetiva_min"] is not None else None,
        ))
    return result
